java/cuvs-java/src/main/java22/com/nvidia/cuvs/internal/TieredIndexImpl.java

```diff
@@ -199,113 +198,123 @@ public SearchResults search(TieredIndexQuery query) throws Throwable {
       try (var resourceAccess = query.getResources().access()) {
         long cuvsRes = resourceAccess.handle();
 
-        long queriesBytes = C_FLOAT_BYTE_SIZE * numQueries * vectorDimension;
-        long neighborsBytes = C_LONG_BYTE_SIZE * numQueries * topK; // 64-bit for tiered index
-        long distancesBytes = C_FLOAT_BYTE_SIZE * numQueries * topK;
-
-        // Allocate DEVICE memory for all data
-        MemorySegment queriesDP = allocateRMMSegment(cuvsRes, queriesBytes);
-        MemorySegment neighborsDP = allocateRMMSegment(cuvsRes, neighborsBytes);
-        MemorySegment distancesDP = allocateRMMSegment(cuvsRes, distancesBytes);
-
-        // Copy queries from host to device
-        var returnValue =
-            cudaMemcpy(queriesDP, hostQueriesSeg, queriesBytes, cudaMemcpyHostToDevice());
-        checkCudaError(returnValue, "cudaMemcpy");
-
-        // Create tensors from device memory
-        long[] queriesShape = {numQueries, vectorDimension};
-        MemorySegment queriesTensor =
-            prepareTensor(localArena, queriesDP, queriesShape, kDLFloat(), 32, kDLCUDA(), 1);
-        long[] neighborsShape = {numQueries, topK};
-        MemorySegment neighborsTensor =
-            prepareTensor(
-                localArena, neighborsDP, neighborsShape, kDLInt(), 64, kDLCUDA(), 1); // 64-bit int
-        long[] distancesShape = {numQueries, topK};
-        MemorySegment distancesTensor =
-            prepareTensor(localArena, distancesDP, distancesShape, kDLFloat(), 32, kDLCUDA(), 1);
-
-        // Sync before prefilter setup
-        returnValue = cuvsStreamSync(cuvsRes);
-        checkCuVSError(returnValue, "cuvsStreamSync");
-
-        // Handle prefilter
-        MemorySegment prefilter = cuvsFilter.allocate(localArena);
-        final MemorySegment prefilterDP;
+        final long queriesBytes = C_FLOAT_BYTE_SIZE * numQueries * vectorDimension;
+        final long neighborsBytes = C_LONG_BYTE_SIZE * numQueries * topK; // 64-bit for tiered index
+        final long distancesBytes = C_FLOAT_BYTE_SIZE * numQueries * topK;
+        final boolean hasPreFilters = query.getPrefilter() != null;
         final long prefilterBytes;
-
-        if (query.getPrefilter() != null) {
+        if (hasPreFilters) {
           BitSet[] prefilters = new BitSet[] {query.getPrefilter()};
```

Comment:
The creation of these arrays here looks redundant, all we need here is to find `prefilterBytes`, which can be 
```
long prefilterDataLength = query.getNumDocs(); // length is 1
long prefilterLen = (prefilterDataLength + 31) / 32;
prefilterBytes = C_INT_BYTE_SIZE * prefilterLen;
```

Reply:
If you want to avoid recomputing them below, you can extract `BitSet[] prefilters` and `long[] prefilterShape` to a final variable here like you did above

Reply:
These are great suggestions.  I'll get to work on these, and update shortly.